File: packages/components/ios/Sources/KiskadeeIOS/Components/Switch/KiskadeeSwitch.swift
import Foundation
import SwiftUI

public struct KiskadeeSwitch: View {
    @Binding private var isOn: Bool

    private let label: String?
    private let theme: KiskadeeTheme
    private let isDisabled: Bool
    private let isInteractionLocked: Bool
    private let interactionCooldown: TimeInterval
    private let showsIcon: Bool

    @State private var isPressed = false
    @State private var dragOffset: CGFloat?
    @State private var gestureStartIsOn: Bool?
    @State private var hasChangedCurrentGesture = false
    @State private var isCoolingDown = false
    @State private var showsActivationFeedback = false
    @State private var cooldownTask: Task<Void, Never>?
    @State private var activationFeedbackTask: Task<Void, Never>?

    public init(
        isOn: Binding<Bool>,
        label: String? = nil,
        theme: KiskadeeTheme,
        isDisabled: Bool = false,
        isInteractionLocked: Bool = false,
        interactionCooldown: TimeInterval = 0,
        showsIcon: Bool = false
    ) {
        self._isOn = isOn
        self.label = label
        self.theme = theme
        self.isDisabled = isDisabled
        self.isInteractionLocked = isInteractionLocked
        self.interactionCooldown = max(0, interactionCooldown)
        self.showsIcon = showsIcon
    }

    public var body: some View {
        switch Result(catching: {
            try KiskadeeSwitchResolver.resolve(
                theme: theme,
                isOn: isOn,
                isPressed: isPressed,
                isDisabled: isDisabled
            )
        }) {
        case let .success(style):
            content(style: style)
        case let .failure(error):
            Text(error.localizedDescription)
                .font(.caption)
                .foregroundStyle(.red)
        }
    }

    private func content(style: KiskadeeSwitchResolvedStyle) -> some View {
        HStack(spacing: 0) {
            if let label {
                Text(label)
                    .font(.system(size: style.labelTextSize ?? 14))
                    .lineSpacing(lineSpacing(style: style))
                    .foregroundStyle(style.labelColor ?? .primary)
                    .padding(.leading, style.labelMarginLeft)
                    .padding(.trailing, style.labelMarginRight)
            }

            switchControl(style: style)
        }
        .fixedSize(horizontal: false, vertical: true)
        .contentShape(Rectangle())
        .gesture(switchGesture(style: style))
        .opacity(isDisabled ? 0.96 : 1)
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(label ?? "Switch")
        .accessibilityValue(isOn ? "On" : "Off")
        .accessibilityAddTraits(isOn ? [.isSelected] : [])
        .accessibilityAction {
            toggle()
        }
    }

    private func switchControl(style: KiskadeeSwitchResolvedStyle) -> some View {
        ZStack(alignment: .topLeading) {
            track(style: style)

            thumbCarrier(style: style)
                .offset(
                    x: style.trackPaddingLeft + thumbOffset(style: style),
                    y: thumbVerticalOffset(style: style)
                )
        }
        .frame(width: style.trackWidth, height: style.trackHeight)
        .animation(.spring(response: 0.22, dampingFraction: 0.86), value: isOn)
        .animation(.easeOut(duration: 0.12), value: isPressed)
        .animation(.easeOut(duration: 0.12), value: showsActivationFeedback)
    }

    private func track(style: KiskadeeSwitchResolvedStyle) -> some View {
        RoundedRectangle(cornerRadius: style.trackRadius, style: .continuous)
            .fill(style.trackColor)
            .overlay {
                RoundedRectangle(cornerRadius: style.trackRadius, style: .continuous)
                    .stroke(style.trackBorderColor, lineWidth: style.trackBorderWidth)
            }
            .frame(width: style.trackWidth, height: style.trackHeight)
    }

    @ViewBuilder
    private func activationFeedbackHalo(
        style: KiskadeeSwitchResolvedStyle,
        thumbWidth: CGFloat,
        thumbHeight: CGFloat,
        thumbRadius: CGFloat
    ) -> some View {
        if let color = style.activationFeedbackColor, style.activationFeedbackSize > 0 {
            let size = style.activationFeedbackSize

            RoundedRectangle(cornerRadius: thumbRadius + size / 2, style: .continuous)
                .stroke(color.opacity(style.activationFeedbackOpacity), lineWidth: size)
                .frame(width: thumbWidth + size, height: thumbHeight + size)
                .opacity(showsActivationFeedback ? 1 : 0)
                .allowsHitTesting(false)
        }
    }

    private func thumbCarrier(style: KiskadeeSwitchResolvedStyle) -> some View {
        let thumbWidth = isOn ? style.thumbWidth : style.thumbRestWidth
        let thumbHeight = isOn ? style.thumbHeight : style.thumbRestHeight
        let radius = min(style.thumbRadius, min(thumbWidth, thumbHeight) / 2)

        return ZStack {
            activationFeedbackHalo(
                style: style,
                thumbWidth: thumbWidth,
                thumbHeight: thumbHeight,
                thumbRadius: radius
            )

            RoundedRectangle(cornerRadius: radius, style: .continuous)
                .fill(style.thumbColor)
                .frame(width: thumbWidth, height: thumbHeight)

            if showsIcon, let iconColor = style.iconColor {
                Image(systemName: isOn ? "checkmark" : "minus")
                    .font(.system(size: iconSize(style: style), weight: .bold))
                    .foregroundStyle(iconColor)
            }
        }
        .frame(width: style.thumbWidth, height: style.thumbHeight)
    }

    private func switchGesture(style: KiskadeeSwitchResolvedStyle) -> some Gesture {
        DragGesture(minimumDistance: 0)
            .onChanged { value in
                guard beginOrContinueGesture() else { return }

                isPressed = true

                let nextOffset = projectedThumbOffset(from: value, style: style)
                dragOffset = nextOffset
                commitIfDragReachedExtreme(nextOffset, style: style)
            }
            .onEnded { value in
                guard !isDisabled && !isInteractionLocked else {
                    resetInteraction()
                    return
                }

                if isTap(value) {
                    _ = commitInteraction(!isOn, startsCooldown: false)
                    resetInteraction(startsCooldown: true)
                    return
                }

                let finalOffset = projectedThumbOffset(from: value, style: style)
                let shouldBeOn = finalOffset >= thumbTravel(style: style) / 2
                _ = commitInteraction(shouldBeOn, startsCooldown: false)
                resetInteraction(startsCooldown: true)
            }
    }

    private func toggle() {
        guard canStartInteraction else { return }

        triggerActivationFeedback()
        _ = commitInteraction(!isOn)
    }

    @discardableResult
    private func commitInteraction(_ nextValue: Bool, startsCooldown: Bool = true) -> Bool {
        guard !isDisabled && !isInteractionLocked else { return false }
        guard nextValue != isOn else { return false }

        if gestureStartIsOn != nil {
            hasChangedCurrentGesture = true
        }

        withAnimation(.spring(response: 0.22, dampingFraction: 0.86)) {
            isOn = nextValue
        }

        if startsCooldown {
            startInteractionCooldown()
        }

        return true
    }

    private func resetInteraction(startsCooldown: Bool = false) {
        let shouldStartCooldown = startsCooldown && hasChangedCurrentGesture

        isPressed = false
        dragOffset = nil
        gestureStartIsOn = nil
        hasChangedCurrentGesture = false

        if shouldStartCooldown {
            startInteractionCooldown()
        }
    }

    private var canStartInteraction: Bool {
        !isDisabled && !isInteractionLocked && !isCoolingDown
    }

    private func beginOrContinueGesture() -> Bool {
        if isDisabled || isInteractionLocked {
            resetInteraction()
            return false
        }

        guard gestureStartIsOn != nil else {
            guard !isCoolingDown else { return false }
            gestureStartIsOn = isOn
            hasChangedCurrentGesture = false
            triggerActivationFeedback()

            return true
        }

        return true
    }

    private func commitIfDragReachedExtreme(_ offset: CGFloat, style: KiskadeeSwitchResolvedStyle) {
        guard gestureStartIsOn != nil else { return }
        let travel = thumbTravel(style: style)
        guard travel > 0 else { return }

        let nextValue: Bool?
        if offset <= 0 {
            nextValue = false
        } else if offset >= travel {
            nextValue = true
        } else {
            nextValue = nil
        }

        guard let nextValue, nextValue != isOn else { return }

        if commitInteraction(nextValue, startsCooldown: false) {
            dragOffset = nextValue ? travel : 0
        }
    }

    private func startInteractionCooldown() {
        cooldownTask?.cancel()

        guard interactionCooldown > 0 else {
            isCoolingDown = false
            return
        }

        isCoolingDown = true
        cooldownTask = Task { @MainActor in
            let nanoseconds = UInt64(interactionCooldown * 1_000_000_000)
            try? await Task.sleep(nanoseconds: nanoseconds)

            if !Task.isCancelled {
                isCoolingDown = false
            }
        }
    }

    private func triggerActivationFeedback() {
        activationFeedbackTask?.cancel()

        showsActivationFeedback = true
        activationFeedbackTask = Task { @MainActor in
            try? await Task.sleep(nanoseconds: 140_000_000)

            if !Task.isCancelled {
                showsActivationFeedback = false
            }
        }
    }

    private func projectedThumbOffset(from value: DragGesture.Value, style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        clampThumbOffset(
            ((gestureStartIsOn ?? isOn) ? thumbTravel(style: style) : 0) + value.translation.width,
            style: style
        )
    }

    private func thumbOffset(style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        dragOffset ?? (isOn ? thumbTravel(style: style) : 0)
    }

    private func thumbTravel(style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        max(0, style.trackWidth - style.trackPaddingLeft - style.trackPaddingRight - style.thumbWidth)
    }

    private func thumbVerticalOffset(style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        let availableHeight = style.trackHeight - style.trackPaddingTop - style.trackPaddingBottom

        return style.trackPaddingTop + max(0, availableHeight - style.thumbHeight) / 2
    }

    private func clampThumbOffset(_ offset: CGFloat, style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        min(max(0, offset), thumbTravel(style: style))
    }

    private func isTap(_ value: DragGesture.Value) -> Bool {
        max(abs(value.translation.width), abs(value.translation.height)) < 1
    }

    private func iconSize(style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        min(style.iconWidth ?? 16, style.iconHeight ?? 16) * 0.72
    }

    private func lineSpacing(style: KiskadeeSwitchResolvedStyle) -> CGFloat {
        guard let textSize = style.labelTextSize,
              let lineHeight = style.labelLineHeight
        else {
            return 0
        }

        return max(0, lineHeight - textSize)
    }
}

public enum KiskadeeSwitchSchemaValidator {
    public static func validate(theme: KiskadeeTheme) throws {
        _ = try KiskadeeSwitchResolver.resolve(
            theme: theme,
            isOn: false,
            isPressed: false,
            isDisabled: false
        )
        _ = try KiskadeeSwitchResolver.resolve(
            theme: theme,
            isOn: true,
            isPressed: false,
            isDisabled: true
        )
    }
}

File: packages/components/android/kiskadee-android/src/main/java/com/kiskadee/android/components/switchcontrol/KiskadeeSwitch.kt
package com.kiskadee.android.components.switchcontrol

import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.snap
import androidx.compose.animation.core.spring
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.input.pointer.changedToUpIgnoreConsumed
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.input.pointer.positionChange
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.onClick
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.stateDescription
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.kiskadee.android.R
import com.kiskadee.android.theme.KiskadeeSwitchResolvedStyle
import com.kiskadee.android.theme.KiskadeeSwitchResolver
import com.kiskadee.android.theme.KiskadeeTheme
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

@Composable
public fun KiskadeeSwitch(
    checked: Boolean,
    onCheckedChange: (Boolean) -> Unit,
    theme: KiskadeeTheme,
    modifier: Modifier = Modifier,
    label: String? = null,
    enabled: Boolean = true,
    showsIcon: Boolean = false,
    interactionLocked: Boolean = false,
    interactionCooldownMillis: Long = 0L,
) {
    var isPressed by remember { mutableStateOf(false) }
    var dragOffset by remember { mutableStateOf<Float?>(null) }
    var gestureStartChecked by remember { mutableStateOf<Boolean?>(null) }
    var gestureChecked by remember { mutableStateOf<Boolean?>(null) }
    var hasChangedCurrentGesture by remember { mutableStateOf(false) }
    var isCoolingDown by remember { mutableStateOf(false) }
    var showsActivationFeedback by remember { mutableStateOf(false) }
    var cooldownJob by remember { mutableStateOf<Job?>(null) }
    var activationFeedbackJob by remember { mutableStateOf<Job?>(null) }

    val scope = rememberCoroutineScope()
    val currentChecked by rememberUpdatedState(checked)
    val currentOnCheckedChange by rememberUpdatedState(onCheckedChange)
    val visualChecked = gestureChecked ?: checked
    val density = LocalDensity.current

    fun currentInteractionValue(): Boolean {
        return gestureChecked ?: currentChecked
    }

    fun startInteractionCooldown() {
        cooldownJob?.cancel()

        if (interactionCooldownMillis <= 0) {
            isCoolingDown = false
            return
        }

        isCoolingDown = true
        cooldownJob = scope.launch {
            delay(interactionCooldownMillis)
            isCoolingDown = false
        }
    }

    fun resetInteraction(startsCooldown: Boolean = false) {
        val shouldStartCooldown = startsCooldown && hasChangedCurrentGesture

        isPressed = false
        dragOffset = null
        gestureStartChecked = null
        gestureChecked = null
        hasChangedCurrentGesture = false

        if (shouldStartCooldown) {
            startInteractionCooldown()
        }
    }

    fun canStartInteraction(): Boolean {
        return enabled && !interactionLocked && !isCoolingDown
    }

    fun triggerActivationFeedback() {
        activationFeedbackJob?.cancel()

        showsActivationFeedback = true
        activationFeedbackJob = scope.launch {
            delay(ACTIVATION_FEEDBACK_DURATION_MILLIS)
            showsActivationFeedback = false
        }
    }

    fun commitInteraction(nextValue: Boolean, startsCooldown: Boolean = true): Boolean {
        if (!enabled || interactionLocked) {
            return false
        }

        if (nextValue == currentInteractionValue()) {
            return false
        }

        if (gestureStartChecked != null) {
            hasChangedCurrentGesture = true
            gestureChecked = nextValue
        }

        currentOnCheckedChange(nextValue)

        if (startsCooldown) {
            startInteractionCooldown()
        }

        return true
    }

    fun toggle() {
        if (!canStartInteraction()) {
            return
        }

        triggerActivationFeedback()
        commitInteraction(!currentInteractionValue())
    }

    fun commitIfDragReachedExtreme(offset: Float, style: KiskadeeSwitchResolvedStyle) {
        if (gestureStartChecked == null) {
            return
        }

        val travel = thumbTravel(style)
        if (travel <= 0f) {
            return
        }

        val nextValue = when {
            offset <= 0f -> false
            offset >= travel -> true
            else -> null
        }

        if (nextValue != null && commitInteraction(nextValue, startsCooldown = false)) {
            dragOffset = if (nextValue) travel else 0f
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            cooldownJob?.cancel()
            activationFeedbackJob?.cancel()
        }
    }

    LaunchedEffect(enabled, interactionLocked) {
        if (!enabled || interactionLocked) {
            resetInteraction()
        }
    }

    LaunchedEffect(interactionCooldownMillis) {
        if (interactionCooldownMillis <= 0) {
            cooldownJob?.cancel()
            isCoolingDown = false
        }
    }

    val resolvedStyle = remember(theme, visualChecked, isPressed, enabled) {
        runCatching {
            KiskadeeSwitchResolver.resolve(
                theme = theme,
                isOn = visualChecked,
                isPressed = isPressed,
                isEnabled = enabled,
            )
        }
    }

    resolvedStyle.fold(
        onSuccess = { style ->
            val densityValue = density.density
            val gestureModifier = Modifier.pointerInput(
                enabled,
                interactionLocked,
                isCoolingDown,
                visualChecked,
                style,
                densityValue,
            ) {
                awaitEachGesture {
                    val down = awaitFirstDown(requireUnconsumed = false)

                    if (!canStartInteraction()) {
                        resetInteraction()

                        var pointerReleased = false
                        while (!pointerReleased) {
                            val event = awaitPointerEvent()
                            pointerReleased = event.changes.any { change ->
                                change.id == down.id && change.changedToUpIgnoreConsumed()
                            }
                        }

                        return@awaitEachGesture
                    }

                    val startChecked = currentInteractionValue()
                    val travel = thumbTravel(style)
                    var totalDragX = 0f
                    var totalDragY = 0f

                    gestureStartChecked = startChecked
                    hasChangedCurrentGesture = false
                    isPressed = true
                    triggerActivationFeedback()

                    while (true) {
                        val event = awaitPointerEvent()
                        val change = event.changes.firstOrNull { it.id == down.id }
                            ?: event.changes.firstOrNull()
                            ?: continue

                        if (change.changedToUpIgnoreConsumed()) {
                            break
                        }

                        val positionChange = change.positionChange()
                        if (positionChange != Offset.Zero) {
                            totalDragX += positionChange.x
                            totalDragY += positionChange.y

                            val nextOffset = clampThumbOffset(
                                (if (startChecked) travel else 0f) + totalDragX / densityValue,
                                style,
                            )

                            dragOffset = nextOffset
                            commitIfDragReachedExtreme(nextOffset, style)
                            change.consume()
                        }
                    }

                    if (!enabled || interactionLocked) {
                        resetInteraction()
                        return@awaitEachGesture
                    }

                    if (max(abs(totalDragX), abs(totalDragY)) < TAP_MOVEMENT_THRESHOLD_PX) {
                        commitInteraction(!currentInteractionValue(), startsCooldown = false)
                        resetInteraction(startsCooldown = true)
                        return@awaitEachGesture
                    }

                    val finalOffset = clampThumbOffset(
                        (if (startChecked) travel else 0f) + totalDragX / densityValue,
                        style,
                    )
                    commitInteraction(finalOffset >= travel / 2f, startsCooldown = false)
                    resetInteraction(startsCooldown = true)
                }
            }

            Row(
                modifier = modifier
                    .semantics(mergeDescendants = true) {
                        role = Role.Switch
                        stateDescription = if (visualChecked) "On" else "Off"

                        if (!enabled) {
                            disabled()
                        }

                        onClick {
                            if (!canStartInteraction()) {
                                false
                            } else {
                                toggle()
                                true
                            }
                        }
                    }
                    .then(gestureModifier),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                if (label != null) {
                    BasicText(
                        text = label,
                        modifier = Modifier.padding(
                            start = style.labelMarginLeft.dp,
                            end = style.labelMarginRight.dp,
                        ),
                        style = TextStyle(
                            color = style.labelColor ?: Color.Unspecified,
                            fontSize = (style.labelTextSize ?: 14f).sp,
                            lineHeight = (style.labelLineHeight ?: style.labelTextSize ?: 20f).sp,
                        ),
                    )
                }

                KiskadeeSwitchControl(
                    checked = visualChecked,
                    dragOffset = dragOffset,
                    style = style,
                    showsIcon = showsIcon,
                    showsActivationFeedback = showsActivationFeedback,
                )
            }
        },
        onFailure = { error ->
            BasicText(
                text = error.message ?: "Invalid Kiskadee switch schema.",
                modifier = modifier,
                style = TextStyle(color = Color.Red, fontSize = 12.sp),
            )
        },
    )
}

@Composable
private fun KiskadeeSwitchControl(
    checked: Boolean,
    dragOffset: Float?,
    style: KiskadeeSwitchResolvedStyle,
    showsIcon: Boolean,
    showsActivationFeedback: Boolean,
) {
    val trackShape = RoundedCornerShape(style.trackRadius.dp)
    val targetThumbOffset = dragOffset ?: if (checked) thumbTravel(style) else 0f
    val animatedThumbOffset by animateDpAsState(
        targetValue = targetThumbOffset.dp,
        animationSpec = if (dragOffset == null) thumbMotionSpec() else snap(),
    )
    val renderedThumbOffset = if (dragOffset == null) animatedThumbOffset else targetThumbOffset.dp

    Box(
        modifier = Modifier.size(width = style.trackWidth.dp, height = style.trackHeight.dp),
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .clip(trackShape)
                .background(style.trackColor)
                .border(
                    border = BorderStroke(style.trackBorderWidth.dp, style.trackBorderColor),
                    shape = trackShape,
                ),
        )

        KiskadeeSwitchThumb(
            checked = checked,
            style = style,
            showsIcon = showsIcon,
            showsActivationFeedback = showsActivationFeedback,
            modifier = Modifier.offset(
                x = style.trackPaddingLeft.dp + renderedThumbOffset,
                y = thumbVerticalOffset(style).dp,
            ),
        )
    }
}

@Composable
private fun KiskadeeSwitchThumb(
    checked: Boolean,
    style: KiskadeeSwitchResolvedStyle,
    showsIcon: Boolean,
    showsActivationFeedback: Boolean,
    modifier: Modifier = Modifier,
) {
    val targetThumbWidth = if (checked) style.thumbWidth else style.thumbRestWidth
    val targetThumbHeight = if (checked) style.thumbHeight else style.thumbRestHeight
    val thumbWidth by animateDpAsState(
        targetValue = targetThumbWidth.dp,
        animationSpec = thumbMotionSpec(),
    )
    val thumbHeight by animateDpAsState(
        targetValue = targetThumbHeight.dp,
        animationSpec = thumbMotionSpec(),
    )
    val thumbRadius = min(style.thumbRadius, min(thumbWidth.value, thumbHeight.value) / 2f).dp

    Box(
        modifier = modifier.size(width = style.thumbWidth.dp, height = style.thumbHeight.dp),
        contentAlignment = Alignment.Center,
    ) {
        KiskadeeSwitchActivationFeedback(
            style = style,
            thumbWidth = thumbWidth,
            thumbHeight = thumbHeight,
            thumbRadius = thumbRadius.value,
            visible = showsActivationFeedback,
        )

        Box(
            modifier = Modifier
                .size(width = thumbWidth, height = thumbHeight)
                .clip(RoundedCornerShape(thumbRadius))
                .background(style.thumbColor),
        )

        if (showsIcon && style.iconColor != null) {
            KiskadeeSwitchIcon(
                checked = checked,
                color = style.iconColor,
                width = style.iconWidth ?: 16f,
                height = style.iconHeight ?: 16f,
            )
        }
    }
}

@Composable
private fun KiskadeeSwitchActivationFeedback(
    style: KiskadeeSwitchResolvedStyle,
    thumbWidth: androidx.compose.ui.unit.Dp,
    thumbHeight: androidx.compose.ui.unit.Dp,
    thumbRadius: Float,
    visible: Boolean,
) {
    val color = style.activationFeedbackColor ?: return
    val size = style.activationFeedbackSize

    if (size <= 0f) {
        return
    }

    val alpha by animateFloatAsState(
        targetValue = if (visible) 1f else 0f,
        animationSpec = spring(
            dampingRatio = Spring.DampingRatioNoBouncy,
            stiffness = Spring.StiffnessMedium,
        ),
    )
    val feedbackShape = RoundedCornerShape((thumbRadius + size / 2f).dp)

    Box(
        modifier = Modifier
            .size(width = thumbWidth + size.dp, height = thumbHeight + size.dp)
            .alpha(alpha)
            .border(
                border = BorderStroke(size.dp, color.copy(alpha = style.activationFeedbackOpacity)),
                shape = feedbackShape,
            ),
    )
}

@Composable
private fun KiskadeeSwitchIcon(
    checked: Boolean,
    color: Color,
    width: Float,
    height: Float,
) {
    Image(
        painter = painterResource(
            id = if (checked) {
                R.drawable.kiskadee_lucide_check
            } else {
                R.drawable.kiskadee_lucide_minus
            },
        ),
        contentDescription = null,
        modifier = Modifier.size(width = width.dp, height = height.dp),
        contentScale = ContentScale.Fit,
        colorFilter = ColorFilter.tint(color),
    )
}

public object KiskadeeSwitchSchemaValidator {
    public fun validate(theme: KiskadeeTheme) {
        KiskadeeSwitchResolver.resolve(
            theme = theme,
            isOn = false,
            isPressed = false,
            isEnabled = true,
        )
        KiskadeeSwitchResolver.resolve(
            theme = theme,
            isOn = true,
            isPressed = false,
            isEnabled = false,
        )
    }
}

private fun thumbTravel(style: KiskadeeSwitchResolvedStyle): Float {
    return max(0f, style.trackWidth - style.trackPaddingLeft - style.trackPaddingRight - style.thumbWidth)
}

private fun thumbVerticalOffset(style: KiskadeeSwitchResolvedStyle): Float {
    val availableHeight = style.trackHeight - style.trackPaddingTop - style.trackPaddingBottom

    return style.trackPaddingTop + max(0f, availableHeight - style.thumbHeight) / 2f
}

private fun clampThumbOffset(offset: Float, style: KiskadeeSwitchResolvedStyle): Float {
    return min(max(0f, offset), thumbTravel(style))
}

private fun <T> thumbMotionSpec() = spring<T>(
    dampingRatio = Spring.DampingRatioNoBouncy,
    stiffness = Spring.StiffnessMedium,
)

private const val ACTIVATION_FEEDBACK_DURATION_MILLIS = 140L
private const val TAP_MOVEMENT_THRESHOLD_PX = 1f
